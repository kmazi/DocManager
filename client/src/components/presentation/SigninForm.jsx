import React from 'react';
import PropTypes from 'prop-types';
import $ from 'jquery';
import Alert from 'sweetalert2';


/**
   * Fires the function that authenticates a user
   * @param {object} event - object containing data about the control that
   * triggered an event
   * @param {object} history - The browser and state history
   * @param {func} signInUser - Function that executes when signing the user in
   * @param {string} roleType - the role in which a user belongs to
   * @param {func} allDocuments - Function that executes when
   * fetching documents belonging to a particular user
   * @return {null} returns void
   */
const signUserIn = (event, history, signInUser, roleType, allDocuments) => {
  // prevent the default behaviour of the button
  event.preventDefault();
  // Get the input values
  const userName = $('#signinform input[type=text]').val();
  const password = $('#signinform input[type=password]').val();
  const formData = { userName, password };
  // consume the returned promise
  signInUser(formData)
    .then((res) => {
      if (res.status === 'successful') {
        allDocuments(roleType);
        history.push('/user/documents');
      } else {
        Alert({
          title: 'Error Signing in',
          text: res.message,
          type: 'error',
          confirmButtonText: 'ok'
        });
      }
    });
};
/**
 * Renders the signin form
 * @param {object} history - Represent the browser and app state history object
 * @param {bool} isAuthentic - Value that shows if user is loggin in
 * @param {func} signInUser - Function that executes when signing in a user
 * @return {object} Returns the signin form to render
 */
const SigninForm = ({ history, signInUser,
  submitButton, roleType, allDocuments }) => (
    <div id="signinform" className="">
      <input type="text" className="forminput" placeholder="Username" />
      <input type="password" className="forminput" placeholder="Password" />
      <button
        id="signin"
        className="center-align waves-effect waves-light btn"
        onClick={event =>
          signUserIn(event, history, signInUser, roleType, allDocuments)}
      >{submitButton}</button>
    </div>
);

SigninForm.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  roleType: PropTypes.string.isRequired,
  signInUser: PropTypes.func.isRequired,
  submitButton: PropTypes.string.isRequired,
  allDocuments: PropTypes.func.isRequired,
};
export default SigninForm;
