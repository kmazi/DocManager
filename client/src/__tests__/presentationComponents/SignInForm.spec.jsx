import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import SigninForm from '../../components/presentation/SigninForm';

describe('The signinform component:', () => {
  const history = { push: url => url };
  const response = 'successful';
  const props = {
    signInUser: jest.fn(() => Promise.resolve(response)),
    allDocuments: jest.fn(() => Promise.resolve()),
    submitButton: 'submit',
    roleType: 'Admin',
    history,
  };
  const enzymeWrapper = mount(<SigninForm {...props} />);
  test('should render the signinform correctly', () => {
    const component = renderer.create(
      <SigninForm {...props} />
  );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test(`should submit the form by firing signInUser when
  it is valid and the submit button is clicked`, () => {
    const event = {
      preventDefault: () => 'hello',
    };
    enzymeWrapper.find('button#signin').props().onClick(event);
    expect(props.signInUser.mock.calls.length).toBe(1);
  });
});
