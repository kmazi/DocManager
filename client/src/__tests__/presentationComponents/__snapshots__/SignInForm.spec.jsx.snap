// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The signinform component: should render the signinform correctly 1`] = `
<div
  className=""
  id="signinform"
>
  <input
    className="forminput"
    placeholder="Username"
    type="text"
  />
  <input
    className="forminput"
    placeholder="Password"
    type="password"
  />
  <button
    className="center-align waves-effect waves-light btn"
    id="signin"
    onClick={[Function]}
  >
    submit
  </button>
</div>
`;
